### DISCUSSION TITLE
Usage examples

### DISCUSSION THREAD

**Discussion by felipeggrod** [sbergen/Responsible#43] [General]
Hey, this framework looks amazing! I am trying to use it to test a multiplayer game, but I'm a bit of a noob with UniRx. So, more usage examples would be great for others in a similar situation trying to use it.

Here are some of my issues. This is an example of a client test, waiting for the server to cycle through some rounds: 

```
[UnityTest]
public IEnumerator TEST_RESPONSIBLE(){
        GameData gameData = gameManager.gameData;
        bool passed = false;

        yield return WaitForCondition("Gather stage entered", () => 
            {
                Debug.Log("gather stage entered");
                return gameData.round.roundStage == RoundStage.gather;
            })
            .AndThen(WaitForCondition("Discuss stage entered", () => gameData.round.roundStage == RoundStage.discuss))
            .AndThen(WaitForCondition("Vote stage entered", () => gameData.round.roundStage == RoundStage.vote))
            .ThenRespondWith("Requirements met", _ => 
            { 
                    Debug.Log("Test Passed. Assert true");
                    passed = true;
            })
            .ExpectWithinSeconds(3);
            //Up until here, everything seems highly intuitive. However there are issues: none of the Debug.Log prints anything on the console.

            //.ContinueWith(Do("oncontinue", onContinue))  // This seem to just call an Action. But if I dont need to call any Action, can I remove this section?
            //.ToYieldInstruction(this.Executor);  // I can't really understand what this does. Where this.Executor comes from? Is this a required call or can I leave it out?

        Debug.Log(passed);
        Assert.IsTrue(passed);
}
```

passed always returns false, the debug.logs dont print anything.
What am I doing wrong?

**Comment by sbergen** [comment]
Thanks for the feedback! The reason you are not getting any of the logs printed is that instead of yielding something that Unity knows how to execute, you are just yielding the test instruction, which Unity does not understand. To fix this, you'll need to construct a `TestInstructionExecutor` instance and use `ToYieldInstruction`. I'd recommend creating a base class that creates a new one in a `SetUp` method. This has to do with some UniRx internals, which I tried to outline here: https://www.beatwaves.net/Responsible/design.html#about-testinstructionexecutor

Regarding
```cs
.ContinueWith(Do("oncontinue", onContinue))  // This seem to just call an Action. But if I dont need to call any Action, can I remove this section?
```

if you don't need to do anything after the previous responder you declared, you don't need this part.

For now, you can look at the test suite for examples: https://github.com/sbergen/Responsible/tree/main/Assets/Responsible/Tests/Runtime
But I'll consider adding some more bare-bones usage examples also.

And sorry for the late reply, I missed this in my notifications, will need to tweak my settings!

**Comment by sbergen** [comment]
@felipeggrod would you mind taking a look at #3 to see if this resolves all your questions? I created a small game and tests for it just to outline how to set things up normally. It also contains improvements in the README. Pleas note that GitHub does not properly support relative links in markdown when in a branch, but the links should work fine once this is merged.

**Comment by felipeggrod** [comment]
@sbergen Thanks! I will try to look at it and get back to you as soon as possible.

**Comment by sbergen** [comment]
Responsible got added to OpenUPM yesterday, and I'll want to get these documentation changes in main ASAP for people discovering it this way. So I'll be merging #3 soon.

@felipeggrod if #3 didn't answer all your questions, feel free to open a new more specific issue for whatever you are missing.

**Comment by sbergen** [comment]
@felipeggrod if you are still interested in Responsible, I just worked on improving the [documentation](https://www.beatwaves.net/Responsible) quite a bit, including an [API reference](https://www.beatwaves.net/Responsible/api/Responsible.html).

Together with the [example tests](https://github.com/sbergen/responsible/tree/main/Packages/Responsible/Samples~/ResponsibleGame/PlayModeTests), I'm hoping it would now be easier for people to get started with Reponsible. But writing easily comprehensible documentation for something you've created yourself can be difficult, so all feedback is welcome!

**Reply by felipeggrod** [reply]
Hey @sbergen, the improvements are looking great!
My project is on hold for a while. I wish I had the time to have a go at it, but I'm busy with life things.
Thanks for the reply!

**Reply by sbergen** [reply]
Life has a tendency to happen :) When you do find some time, don't hesitate to ask, if you face any issues!
